src/coreclr/jit/llvm.cpp

```diff
@@ -2698,32 +2774,75 @@ void Llvm::lowerToShadowStack()
     }
 }
 
+
 //------------------------------------------------------------------------
 // Convert GT_STORE_LCL_VAR and GT_LCL_VAR to use the shadow stack when the local needs to be GC tracked,
 // rewrite calls that returns GC types to do so via a store to a passed in address on the shadow stack.
 // Likewise, store the returned value there if required.
 //
-void Llvm::PlaceAndConvertShadowStackLocals()
+void Llvm::Lower()
 {
     populateLlvmArgNums();
 
     _shadowStackLocalsSize = 0;
 
     std::vector<LclVarDsc*> locals;
     unsigned localsParamCount = 0;
+    unsigned lvaCount = _compiler->lvaCount;
 
-    for (unsigned lclNum = 0; lclNum < _compiler->lvaCount; lclNum++)
+    for (unsigned lclNum = 0; lclNum < lvaCount; lclNum++)
     {
         LclVarDsc* varDsc = _compiler->lvaGetDesc(lclNum);
         if (!canStoreLocalOnLlvmStack(varDsc))
         {
             locals.push_back(varDsc);
             if (varDsc->lvIsParam)
             {
+                if (varDsc->lvPromotedStruct())
+                {
+                    // TODO-LLVM: the offsets for promoted structs on the shadow stack is calculated incorrectly below in the call to SetStackOffset
```

Comment:
> the offsets for promoted structs on the shadow stack is calculated incorrectly below in the call to SetStackOffset

The fix for this is not exactly simple, but it will make us better at not placing more than necessary on the shadow stack.

Cases:
1) Independent promotion:
  - Locals: we need to place the GC fields on the shadow stack, the rest can be ordinary locals. The parent should be skipped.
  - Args: these we will rewrite into ordinary locals below, from which the fields, which can be both GC and non-GC will be initialized. Thus, both the parent and all its GC fields need to have shadow stack offsets.
2) Dependent promotion: field references will be folded into the parent, no need to assign shadow stack offsets for them (though we could: `parentOffset + lvFieldOffset`).

Reply:
commit https://github.com/dotnet/runtimelab/pull/2007/commits/e3de1e1865a07e092ef803c0afacaefb8ec5d487 is my attempt at point 2 here, the dependent promotion

Reply:
commit https://github.com/dotnet/runtimelab/pull/2007/commits/204bbdc726f3d038ac2a8bd845980e63766452e3 is a first attempt at the independent nongc field.  However I have a couple of doubts.  First I never seems to get any fields with an offset other than 0, aren't we expecting to get some?  Secondly, I suppose I've got this code in the wrong place because for the gc fields I'm going to want the shadowstack offsets, so I'll have to move this to `Lower` and use `Uses` rather than walking the IR tree?

Reply:
There are two parts to the independent promotion logic.

1) Shadow stack layout. Essentially, we should exclude parents for the independent case and include fields for the dependent case (as outlined above). Probably the easiest to reason about way to do this is to move the `locals.push_back(varDsc);` code _after_ the code that deals with promotion, it should look like:
```c
        if (!canStoreLocalOnLlvmStack(varDsc))
        {
            if (lvaGetPromotionType(varDsc) == PROMOTION_TYPE_INDEPENDENT)
            {
                // The individual fields will placed on the shadow stack.
                continue;
            }
            if (lvaIsFieldOfDependentlyPromotedStruct(varDsc))
            {
                // The fields will be referenced through the parent.
                continue;
            }

            locals.push_back(varDsc);
            if (varDsc->lvIsParam)
            {
                localsParamCount++;
            }
        }
```
With this, and the reordering, I believe the shadow stack offsets should be correct.

2. "Fake" independent promotion. There is one (prominent, there are also some corner cases) case that we need to solve where an independently promoted struct is still actually referenced in the IR: `CanBeReplacedWithItsField` structs. They will appear as `STORE_LCL_VAR<parent>(CALL<struct>)` in IR and need to be rewritten to refer to the individual field instead (this lowering, naturally, should take place before on-shadow-stack locals are rewritten). The easiest way to do this would be to turn the store into `STORE_LCL_FLD<field>(CALL<struct>)`, alas, we need to wait for some upstream updates for that, so `STORE_OBJ(LCL_VAR_ADDR<field>, CALL<struct>)` will have to do.

Reply:
Thanks a lot for the code.  I think I'm down to the `CanBeReplacedWithItsField` case down.

Reply:
I have this case
```
N005 ( 22,  6) [000076] --CXG-------        t76 = *  CALL      struct System.DateTimeOffset.ValidateDate $c4
                                                  /--*  t76    struct
N007 ( 26,  9) [000085] DA-XG-------              *  STORE_LCL_VAR struct<System.DateTime, 8>(P) V06 tmp6         d:1
                                                  *    long   V06._dateData (offs=0x00) -> V23 tmp23
```
Where `CanBeReplacedWithItsField` is true for `V06` at node 85.  Is this one of the corner cases and can if I handle this the same as the  `STORE_LCL_VAR<parent>(CALL<struct>)` case would it be a problem?  

This is `S_P_CoreLib_System_DateTimeOffset___cctor`,  full log attached.
[CanBeReplacedWithItsField.txt](https://github.com/dotnet/runtimelab/files/9944775/CanBeReplacedWithItsField.txt)

Reply:
It is the `STORE_LCL_VAR<parent>(CALL<struct>)` case itself (`parent` would be the promoted `V06` here).

Reply:
Thanks,

I have this code
```
            unsigned storeLclNum  = storeLclNode->GetLclNum();
            LclVarDsc* fieldDsc   = _compiler->lvaGetDesc(addrVarDsc->lvFieldLclStart + 0);
            ClassLayout* layout   = addrVarDsc->GetLayout();
            GenTree* nodeToStore  = storeLclNode->gtGetOp1();
            var_types addrVarType = addrVarDsc->TypeGet();

            storeLclNode->SetOper(GT_LCL_FLD_ADDR);
            storeLclNode->AsLclFld()->SetLclOffs(0);
            storeLclNode->ChangeType(fieldDsc->TypeGet());
            // TODO-LLVM: should this be set?
            // storeLclNode->AsLclFld()->SetFieldSeq(nullptr);

            unsigned fldAddrTempNum = _compiler->lvaGrabTemp(true DEBUGARG("field address"));
            storeLclNode->SetLclNum(fldAddrTempNum);

            GenTree* storeObjNode = createStoreNode(addrVarType, storeLclNode, nodeToStore, layout);
            GenTree* lclNode      = _compiler->gtNewLclFldNode(storeLclNum, fieldDsc->TypeGet(), 0);

            CurrentRange().InsertAfter(storeLclNode, storeObjNode, lclNode);
```
I change the `GT_STORE_LCL_VAR` to a `GT_LCL_FLD_ADDR`  and a store followed by a `gtNewLclFldNode` so I can get the value in the lclvar that was originally the destination of the `STORE_LCL_VAR`  .  Do I need to call `SetFieldSeq`, I'm guessing I do as weird stuff seems to happen if I dont?

Reply:
>  Do I need to call `SetFieldSeq` ?

Yes, but it's simpler to just do `SetOper(GT_LCL_VAR_ADDR);`.

Also, the local address node should be `TYP_I_IMPL`, not `fieldDsc->TypeGet()`.

I don't understand the need for `fldAddrTempNum` and `lclNode` here, the whole transformation is just this:
```scala
t1 = CALL
     STORE_LCL_VAR(t1) V00 (V06 field)
```
=>
```scala
t1 = CALL
t2 = LCL_VAR_ADDR V06
     STORE_OBJ(t2, t1)
```

Reply:
Thanks, should `V06` in your example participate in SSA renaming?  I have an assert in my example above where `V23` does not evaluate true for
```
       if (varDsc->lvIsParam || m_pCompiler->info.compInitMem || varDsc->lvMustInit ||
            VarSetOps::IsMember(m_pCompiler, m_pCompiler->fgFirstBB->bbLiveIn, varDsc->lvVarIndex))
```
and is not pushed to the rename stack and fails in the pop.
https://github.com/dotnet/runtimelab/blob/780dfdc76d9d0a2083e98e367a85c7bc1ed18c51/src/coreclr/jit/ssabuilder.cpp#L1581-L1590

https://github.com/dotnet/runtimelab/blob/780dfdc76d9d0a2083e98e367a85c7bc1ed18c51/src/coreclr/jit/ssarenamestate.cpp#L51

Reply:
> should `V06` in your example participate in SSA renaming

Presumably not, since us taking its address means it should be marked as `lvHasLocalAddr`.

The best way to ensure that happens is to make the lowering function return "the next node to lower", so after the store is transformed both the `LCL_VAR_ADDR` and `STORE_OBJ` are lowered.

Reply:
Thanks, l moved the `lowerStoreLcl` to nearly the top of the method to avoid messing with the node iterator.  Looks like it works, so maybe we can leave it like that?